File: test/datasource.cpp
#include <catch2/catch_test_macros.hpp>

#include "mapget/datasource/datasource.h"
#include "httplib.h"
#include "mapget/model/stream.h"

using namespace mapget;

TEST_CASE("DataSource", "[DataSource]")
{
    // Create DataSourceInfo
    auto info = DataSourceInfo::fromJson(R"(
    {
        "mapId": "GarlicChickenMap",
        "layers": {
            "WayLayer": {
                "featureTypes":
                [
                    {
                        "name": "Way",
                        "uniqueIdCompositions":
                        [
                            [
                                {
                                    "partId": "areaId",
                                    "description": "String which identifies the map area.",
                                    "datatype": "STR"
                                },
                                {
                                    "partId": "wayId",
                                    "description": "Globally Unique 32b integer.",
                                    "datatype": "U32"
                                }
                            ]
                        ]
                    }
                ]
            }
        }
    }
    )"_json);
    // configure info

    // Initialize a DataSource
    DataSource ds(info);

    ds.onTileRequest([](TileFeatureLayer& tile) {
        auto f = tile.newFeature("Way", {{"areaId", "Area42"}, {"wayId", 0}});
        auto g = f->geom()->newGeometry(GeomType::Line);
        g->append({42., 11});
        g->append({42., 12});
    });

    // Launch the DataSource on a separate thread
    ds.go();

    // Ensure the DataSource is running
    REQUIRE(ds.isRunning() == true);

    SECTION("Fetch /info")
    {
        // Initialize an httplib client
        httplib::Client cli("localhost", ds.port());

        // Send a GET info request
        auto fetchedInfoJson = cli.Get("/info");
        auto fetchedInfo = DataSourceInfo::fromJson(nlohmann::json::parse(fetchedInfoJson->body));
        REQUIRE(fetchedInfo.toJson() == info.toJson());
    }

    SECTION("Fetch /tile")
    {
        // Initialize an httplib client
        httplib::Client cli("localhost", ds.port());

        // Send a GET tile request
        auto tileResponse = cli.Get("/tile?layer=WayLayer&tileId=1");

        // Check that the response is OK
        REQUIRE(tileResponse != nullptr);
        REQUIRE(tileResponse->status == 200);

        // Check the response body for expected content
        auto receivedTileCount = 0;
        TileLayerStream::Reader reader(
            [&](auto&& mapId, auto&& layerId)
            {
                REQUIRE(mapId == info.mapId_);
                return info.layers_[std::string(layerId)];
            },
            [&](auto&& tile) { receivedTileCount++; });
        reader.read({tileResponse->body.begin(), tileResponse->body.end()});

        REQUIRE(receivedTileCount == 1);
    }

    SECTION("Stop the data source")
    {
        ds.stop();
        REQUIRE(ds.isRunning() == false);
    }

    SECTION("Wait for data source")
    {
        auto waitThread = std::thread([&]{ds.waitForSignal();});
        ds.stop();
        waitThread.join();
        REQUIRE(ds.isRunning() == false);
    }
}

File: examples/python/datasource.py
#!python

import mapget
from ndslive.math import PackedTileId
from mapget import Point as Pt
import sys


def handle_tile_request(tile: mapget.TileFeatureLayer):
    # Read out requested tile-id / map-id / layer-id
    print(f"Got request for tile={tile.tile_id().value:02X}, map={tile.map_id()}, layer={tile.layer_id()}.")

    # Create a new feature
    feature: mapget.Feature = tile.new_feature("Way", [("wayId", 0)])

    # Assign geometry (low-level)
    geom: mapget.Geometry = feature.geom().new_geometry(mapget.GeomType.LINE)
    geom.append(41., 11.)
    geom.append(Pt(x=42., y=12., z=506))

    # Assign geometry (high-level)
    feature.add_point(Pt(42.5, 11.6))
    feature.add_points([Pt(42.5, 11.6), Pt(42.5, 11.8)])
    feature.add_line([Pt(42.5, 11.6), Pt(42.5, 11.7)])
    feature.add_mesh([Pt(42.5, 11.6), Pt(42.5, 11.7), Pt(42.2, 11.7)])
    feature.add_poly([Pt(42.5, 11.6), Pt(42.5, 11.7), Pt(42.2, 11.7), Pt(42.2, 11.3)])

    # Add an attribute
    fixed_attrs: mapget.Object = feature.attributes()
    fixed_attrs.add_field("isBridge", False)

    # Add an attribute which has a compound value (low-level)
    attr_obj = tile.new_object()
    attr_obj.add_field("name", "Main St.")
    attr_obj.add_field("houseNumber", 5)
    attr_arr = tile.new_array()
    attr_arr.append(attr_obj)
    attr_arr.append(attr_obj)
    fixed_attrs.add_field("addresses", attr_arr)

    # Add an attribute which has a compound value (high-level)
    # Note: Map values may also be feature IDs to create references.
    fixed_attrs.add_field("pois", [
        {"name": "Bakery", "rating": 10},
        {"name": "Gas Station", "rating": 8},
        {"reference": tile.new_feature_id("Way", [("wayId", 7)])}
    ])

    # Add an attribute layer
    attr_layer: mapget.Object = feature.attribute_layers().new_layer("rules")
    attr: mapget.Attribute = attr_layer.new_attribute("SPEED_LIMIT")
    attr.validity().new_offset_range(
        mapget.ValidityGeometryOffsetType.RELATIVE_LENGTH,
        0.0,
        1.0,
        direction=mapget.Direction.POSITIVE)
    attr.add_field("speedLimit", 50)

    # Add a feature relation
    target = tile.new_feature_id("Way", [("wayId", 10)])
    relation = feature.add_relation("successor", target)
    relation.source_validity().new_complete()

    # Attach source-data provenance.
    refs = tile.new_source_data_references([
        ("RawWayLayer", "primary", mapget.SourceDataAddress(0, 128))
    ])
    feature.set_source_data_references(refs)


def handle_source_data_request(tile: mapget.TileSourceDataLayer):
    compound = tile.new_compound(2)
    compound.set_schema_name("example.RawWay")
    compound.set_source_data_address(mapget.SourceDataAddress(0, 128))
    compound.add_field("wayId", 0)
    compound.add_field("name", "Main St.")
    tile.add_root(compound)


def handle_locate_request(request: mapget.LocateRequest):
    response = mapget.LocateResponse(request)
    response.tile_key = mapget.MapTileKey(
        mapget.LayerType.FEATURES,
        request.map_id,
        "WayLayer",
        PackedTileId.from_tile_xy(0, 0, 0),
        0)
    return [response]


def handle_cache_expired(tile_key: mapget.MapTileKey, expired_at_us: int):
    print(f"Cached tile expired: {tile_key} at {expired_at_us}us.")


# Instantiate a data source with a minimal mandatory set
# of meta-information.
ds = mapget.DataSourceServer({
    "layers": {
        "WayLayer": {
            "featureTypes": [
                {
                    "name": "Way",
                    "uniqueIdCompositions": [[{
                        "partId": "wayId",
                        "datatype": "I64"
                    }]]
                }
            ]
        },
        "RawWayLayer": {
            "type": "SourceData"
        }
    }, "mapId": "TestMap"
})

# Set the callback which is invoked when a tile is requested.
ds.on_tile_feature_request(handle_tile_request)
ds.on_tile_sourcedata_request(handle_source_data_request)
ds.on_locate_request(handle_locate_request)
ds.on_cache_expired(handle_cache_expired)

# Parse port as optional first argument
port = 0  # Pick random free port
if len(sys.argv) > 1:
    port = int(sys.argv[1])

# Run the data source - you may also set port=0 to select a
# port automatically.
ds.go(port=port)

# Wait until Ctrl-C is hit. Navigate e.g. to
#  http://localhost:54544/tile?layer=WayLayer&tileId=2&responseType=json
# to test the running data source.
ds.wait_for_signal()
